src/balanced.jl

```diff
@@ -76,12 +78,16 @@ img_binary = binarize(Balanced(), img)
 
 1. “BI-LEVEL IMAGE THRESHOLDING - A Fast Method”, Proceedings of the First International Conference on Bio-inspired Systems and Signal Processing, 2008. Available: [10.5220/0001064300700076](https://doi.org/10.5220/0001064300700076)
 """
-function binarize(algorithm::Balanced,  img::AbstractArray{T,2}) where T <: Colorant
-    img₀₁ = zeros(Gray{Bool}, axes(img))
+struct Balanced <: AbstractImageBinarizationAlgorithm end
+
+function (f::Balanced)(out::GenericGrayImage, img::GenericGrayImage)
     edges, counts = build_histogram(img,  256)
     t = find_threshold(HistogramThresholding.Balanced(), counts[1:end], edges)
     for i in CartesianIndices(img)
-      img₀₁[i] = img[i] < t ? 0 : 1
+      out[i] = img[i] < t ? 0 : 1
     end
-    img₀₁
+    out
 end
+
+(f::Balanced)(out::GenericGrayImage, img::AbstractArray{<:Color3}) =
+    f(out, of_eltype(Gray, img))
```

Comment:
If remove this, the test would fail on RGB images. Looks like `build_histogram` doesn't support `Color3` images yet?

```diff
diff --git a/src/balanced.jl b/src/balanced.jl
index a72291b..52a6ba9 100644
--- a/src/balanced.jl
+++ b/src/balanced.jl
@@ -80,7 +80,8 @@ img_binary = binarize(Balanced(), img)
 """
 struct Balanced <: AbstractImageBinarizationAlgorithm end

-function (f::Balanced)(out::GenericGrayImage, img::GenericGrayImage)
+function (f::Balanced)(out::GenericGrayImage,
+                       img::AbstractArray{T, 2}) where T<:Union{Number, Colorant}
     edges, counts = build_histogram(img,  256)
     t = find_threshold(HistogramThresholding.Balanced(), counts[1:end], edges)
     for i in CartesianIndices(img)
@@ -88,6 +89,3 @@ function (f::Balanced)(out::GenericGrayImage, img::GenericGrayImage)
     end
     out
 end
-
-(f::Balanced)(out::GenericGrayImage, img::AbstractArray{<:Color3}) =
-    f(out, of_eltype(Gray, img))
```

Reply:
Currently the `build_histogram` method converts an image to a single channel (Gray) before it computes the histogram. If one wants to compute the histogram for an RGB image one would need to call `build_histogram` for each channel separately. 

It looks like I forgot to make sure that the image is converted to Grayscale when the user specifies the minval option. For example, I currently have

```julia
function build_histogram(img::AbstractArray, nbins::Integer = 256)
    build_histogram(img, nbins, minval =  minfinite(img), maxval = maxfinite(img))
end

function build_histogram(img::AbstractArray{T}, edges::AbstractRange) where {T<:Color3}
    build_histogram(Gray.(img),edges)
end
```

I forgot to add something like
```julia
function build_histogram(img::AbstractArray, nbins::Integer = 256) where {T<:Color3}
    img_gray = Gray.(img)
    build_histogram(img_gray, nbins, minval =  minfinite(img_gray), maxval = maxfinite(img_gray))
end
```